### PR TITLE
SW-4310 Allow replacing observation plots that aren't completed from the results table, intermediate implementation

### PR DESCRIPTION
**Pull request by karthikbtf**
- PR is built on top off the redux/service [PR](https://github.com/terraware/terraware-web/pull/1729) which also requires a couple BE PRs to land, the BE effort is WIP
- Added an opt-in flag to expose this feature
- Refactored TableRowPopupMenu for reusability
- Added popup menu and modal in the zone view of observation results
- only monitoring plots that aren't completed yet will have the menu enabled
- this implementation is an intermediate iteration (does make the API request)
- processing of results and display success message description will be a follow up PR (waiting on UX clarification, etc.)

Terraware App 2023-10-16 16-42-52

Terraware App 2023-10-16 16-42-27

Terraware App 2023-10-16 16-43-13

### CODE REVIEW

File: src/components/Observations/zone/index.tsx [hunk 1]
```diff
@@ -95,34 +125,47 @@ export default function ObservationPlantingZone(): JSX.Element {
...
+                    setReplaceObservationPlot
+                  )}
+                />
+              </Box>
+            </Card>
+          </Grid>
         </Grid>
-      </Grid>
-    </DetailsPage>
+      </DetailsPage>
```
Comment (karthikbtf) [inline src/components/Observations/zone/index.tsx:168]: all of this was just indented 2 spaces

File: src/components/common/table/TableRowPopupMenu.tsx [hunk 2]
```diff
@@ -1,19 +1,20 @@
...
+  label: string;
+  onClick: () => void;
 };
 
-const ActionsMenu = ({ observationId, state }: ActionsMenuProps): JSX.Element => {
+type TableRowPopupMenuProps = {
+  menuItems: TableRowPopupMenuItem[];
+};
+
+const TableRowPopupMenu = ({ menuItems }: TableRowPopupMenuProps): JSX.Element => {
```
Comment (karthikbtf) [inline src/components/common/table/TableRowPopupMenu.tsx:16]: This was refactored/renamed for reusability.

File: src/types/Observations.ts [hunk 3]
```diff
@@ -100,17 +100,6 @@ export const getPlotStatus = (status?: MonitoringPlotStatus): string => {
   }
 };
 
-export const getReplaceObservationPlotDuration = (duration: ReplaceObservationPlotDuration): string => {
```
Comment (karthikbtf) [inline src/types/Observations.ts:103]: this was added in the parent PR but really not needed

File: src/components/Observations/zone/ReplaceObservationPlotModal.tsx [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  const snackbar = useSnackbar();
+  const theme = useTheme();
+  const [justification, setJustification] = useState<string>();
+  const [duration, setDuration] = useState<ReplaceObservationPlotDuration | undefined>();
+  const [validate, setValidate] = useState<boolean>(false);
+  const [requestId, setRequestId] = useState<string>('');
+
+  const result = useAppSelector((state) => selectReplaceObservationPlot(state, requestId));
+
+  const replaceObservationPlot = () => {
```
Comment (charlie-tf) [inline src/components/Observations/zone/ReplaceObservationPlotModal.tsx:33]: Does this need to call onClose?

**Review by charlie-tf — APPROVED** [review]
Looks pretty good, just a minor question

File: src/components/Observations/zone/ReplaceObservationPlotModal.tsx (hunk 4 above)
Comment (karthikbtf) [reply src/components/Observations/zone/ReplaceObservationPlotModal.tsx:33]: Thanks for the review Charlie. `replaceObservationPlot` function dispatches the request.
We have a listener on the result selector, which calls `onClose` if the request succeeded.
This modal handles the entire request/response handling.
Will be reused in the list view and the map view (next effort).

**Pull request merged**